API_GUIDE.md

```diff
@@ -1,76 +1,118 @@
-# PyTorch/XLA API And Best Practices
+# PyTorch on XLA Devices
 
-## XLA Tensors
+PyTorch runs on XLA devices, like Cloud TPUs, with the [torch_xla package](https://github.com/pytorch/xla/). This document describes how to run your models on these devices.
 
-PyTorch/XLA adds a new device, similarly to CPU and GPU devices. The following snippet creates an XLA tensor filled with random values, then prints the device and the contents of the tensor:
+## Creating an XLA Tensor
+
+PyTorch/XLA adds a new `xla` device type to PyTorch. This device type works just
+like other PyTorch device types. For example, here's how to create and
+print an XLA tensor:
 
 ```python
 import torch
 import torch_xla
 import torch_xla.core.xla_model as xm
 
-x = torch.randn(4, 2, device=xm.xla_device())
-print(x.device)
-print(x)
+t = torch.randn(2, 2, device=xm.xla_device())
+print(t.device)
+print(t)
+```
+
+This code should look familiar. PyTorch/XLA uses the same interface as regular
+PyTorch with a few additions. Importing `torch_xla` initializes PyTorch/XLA, and
+`xm.xla_device()` returns the current XLA device. This may be a CPU or Cloud TPU
+depending on your environment.
+
+## XLA Tensors are PyTorch Tensors
+
+PyTorch operations can be performed on XLA tensors just like CPU or CUDA tensors.
+
+For example, XLA tensors can be added together:
+
+```python
+t0 = torch.randn(2, 2, device=xm.xla_device())
+t1 = torch.randn(2, 2, device=xm.xla_device())
+print(t0 + t1)
+```
+
+Or matrix multiplied:
+
+```python
+print(t0.mm(t1))
+```
+
+Or used with neural network modules:
+
+```python
+l_in = torch.randn(10, device=xm.xla_device())
+linear = torch.nn.Linear(10, 20).to(xm.xla_device())
+l_out = linear(l_in)
+print(l_out)
 ```
 
-The XLA device is not a physical device but instead stands in for either a Cloud TPU or CPU. The underlying storage for XLA tensors is a contiguous buffer in device memory and the code in the model shouldn't assume any stride.
+Like other device types, XLA tensors only work with other XLA tensors on the
+same device. So code like
 
-XLA Tensor doesn't support converting single tensor to half precision using `tensor.half()`. Instead, environment variable `XLA_USE_BF16` is available, which converts **all** PyTorch float values to bfloat16 when sending them to the TPU device. The conversion is totally transparent to the user, and the XLA tensors will still retain a float dtype. Similarly, when the tensor is moved back to CPU, its type will be float.
+```python
+l_in = torch.randn(10, device=xm.xla_device())
+linear = torch.nn.Linear(10, 20).cpu()
```

Comment:
I think you can omit `.cpu()` here as it is the default.

Reply:
Sure.